### DISCUSSION TITLE
HMR issue

### DISCUSSION THREAD

**Discussion by abelmark** [webpack/webpack#15074] [Q&A]
# Bug report

**What is the current behavior?**
I am trying to instantiate HMR with a React app through webpack-dev-server.  My app is run on webpack-dev-server and the assets are being loaded through a docker container running a PHP slim application being read from an `assets-manifest.json`.  When I update my code and save I can see webpack-dev-server is aware due to logs showing up in the browser console, but it ends up failing when trying to load `hot-update.json`

**If the current behavior is a bug, please provide the steps to reproduce.**
1.  Boot up webpack-dev-server
2. Make changes to React app and save
3. webpack-dev-server and HMR updates are logged into the browser console
4. Changes fail when trying to load `hot-module.json` as the URL is missing a `/` between the host and the file `http://docker.for.mac.localhost:9044main.b84834ae07a0f0a94465.hot-update.json`

Browser console logging:
```
[webpack-dev-server] App updated. Recompiling...
index.js?e2a7:551 
[webpack-dev-server] App hot update...
log.js?7596:24 
[HMR] Checking for updates on the server...
react_devtools_backend.js:4045 
[HMR] Update failed: TypeError: Failed to execute 'fetch' on 'Window': Failed to parse URL from http://docker.for.mac.localhost:9
044main.b84834ae07a0f0a94465.hot-update.json
```

My current `devServer` config:

```
  devServer: {
    host: '0.0.0.0',
    allowedHosts: ['docker.for.mac.localhost'],
    port: 9044,
  },
```

**What is the expected behavior?**
The correct URL should be hit for the `hot-update.json` (ie including a `/` in the url of  `http://docker.for.mac.localhost:9044/main.b84834ae07a0f0a94465.hot-update.json`)HMR should update the page without reloading and [HMR] should not log an error

**Other relevant information:**
webpack version:  Webpack 5.65.0
Node.js version: 14.17.6
Operating System: macOS
Additional tools: Docker, PHP Slim, React

**Comment by alexander-akait** [comment]
Can you provide configuration, I think problem in `output.publicPath`, you need `/` at the end

**Comment by abelmark** [comment]
For my `publicPath` I had tried `''`, `auto` and `/` and none of these seem to resolve the issue.  When I add a `/` my application won't even load and I get this error since there is an extra `/`:

```
GET http://docker.for.mac.localhost:9044//main.bundle.min.js?version=200 404 (Not Found)
042efeb6-e001-4b5c-8724-d53da7a01f8d:17 
GET http://docker.for.mac.localhost:9044//vendor.bundle.min.js?version=200 net::ERR_ABORTED 404 (Not Found)
042efeb6-e001-4b5c-8724-d53da7a01f8d:16 
GET http://docker.for.mac.localhost:9044//main.bundle.min.js?version=200 net::ERR_ABORTED 404 (Not Found)
042efeb6-e001-4b5c-8724-d53da7a01f8d:17 
GET http://docker.for.mac.localhost:9044//vendor.bundle.min.js?version=200 net::ERR_ABORTED 404 (Not Found)
```

Here is my full output config for this issue, prior I had it on `auto`:

```
  output: {
    filename: `[name].bundle.min.js?version=${version}`,
    path: path.resolve(__dirname, 'dist'),
    chunkFilename: '[name].[chunkhash].min.js',
    publicPath: '/',
  },
```

**Reply by alexander-akait** [reply]
Check you proxy configuration, looks like you already added `/`

**Reply by abelmark** [reply]
I posted my `devServer` config above.  I don't have any proxy settings,

```
  devServer: {
    host: '0.0.0.0',
    allowedHosts: ['docker.for.mac.localhost'],
    port: 9044,
  },
```

webpack-dev-server is serving my app which is being called through an asset-manifest by a PHP application run in a docker container.

**Reply by alexander-akait** [reply]
How you proxy requests to php application?

**Reply by abelmark** [reply]
The PHP application retrieves the React assets through an
asset-manifest.json. We hit the php application port in the url so no proxy
needed.

**Reply by alexander-akait** [reply]
Please provide reproducible example and I will should how to fix it, we don't add additional `/`, if you set `publicPath` to `/`, we will use only this for public path, it is just string, maybe `asset-manifest.json` contains `/` so you apply `/` twice?

**Reply by abelmark** [reply]
So I figured out why it was applying two `/`.  Our php app was injecting the assets like so 

```
<?php foreach ($resource_assets['js'] as $jsFile) { ?>
  <script type="text/javascript" src="<?php print $sdlc_assets . '/' . $jsFile; ?>"></script>
<?php } ?>
```

I removed the extra `/` and changed the `publicPath` in output to:

```
publicPath: '/`
```

and I no longer have the double `/` issue.  However, HMR on the devServer is still negating the `/` when trying to load the js after an update:

```
react_devtools_backend.js:4045 [HMR] Update failed: TypeError: Failed to execute 'fetch' on 'Window': Failed to parse URL from http://docker.for.mac.localhost:9044main.18e8d82354b08e16ef09.hot-update.json
```

devServer config:

```
  devServer: {
    allowedHosts: ['docker.for.mac.localhost'],
    port: 9044,
    watchContentBase: true,
  },
  ```

**Reply by abelmark** [reply]
Still seeing this issue.  The HMR URL is missing the '/'

**Reply by alexander-akait** [reply]
Check your proxy, again - we are generating URL as you provided a configuration, if something was removed/not generated check it on an application side, also avoid to set `publicPath`, by default webpack generates relative URLs

**Reply by alexander-akait** [reply]
Check `filename`/`chunkFilename` too and avoid having `/ at the start

**Reply by abelmark** [reply]
When HMR tries to load the hot update I see this warning in the console

`[HMR] Update failed: TypeError: Failed to execute 'fetch' on 'Window': Failed to parse URL from http://docker.for.mac.localhost:9044main.fda380b71f3b021dc4e9.hot-update.json
    at __webpack_require__.hmrM (http://docker.for.mac.localhost:9044/main.bundle.min.js?version=200:6432:20)`
    
If you look closely you can see that the '/' is missing between the port and the file name

**actual:** http://docker.for.mac.localhost:9044main.fda380b71f3b021dc4e9.hot-update.json
**expected:** http://docker.for.mac.localhost:9044/main.fda380b71f3b021dc4e9.hot-update.json
    
My app architecture is I run webpack dev server on port 9044 to host my UI assets.  My main application is run through a PHP application hosted in a docker container that reads the assets from an `asset-manifest.json`.  Here is my dev server config for development

```javascript
devServer: {
  allowedHosts: ['docker.for.mac.localhost'],
  port: 9044,
},
```

webpack-config.js
```javascript
entry: ['@babel/polyfill', './src/index.js'],
output: {
  filename: `[name].bundle.min.js?version=${version}`,
  path: path.resolve(__dirname, 'dist'),
  publicPath: '/',
},
```

my AssetManifestPlugin

```javascript
new WebpackAssetsManifest({
  publicPath: true,
  writeToDisk: true,
  output: './asset-manifest.json',
  merge: true,
}),
```

asset-manifest.json output
```json
{
  "favicon.ico": "/favicon.ico",
  "main.js": "/main.bundle.min.js?version=200",
  "vendor.js": "/vendor.bundle.min.js?version=200"
}
```

**Comment by abelmark** [comment]
I'll check again but as I mentioned we aren't proxying any requests. We're
generating an asset-manifest and serving our app through a PHP Backend.
